File: Android/app/src/main/java/com/apps/kruszyn/lightorganapp/ui/NumberPickerPreference.java
package com.apps.kruszyn.lightorganapp.ui;

import android.content.Context;
import android.content.res.TypedArray;
import android.os.Parcel;
import android.os.Parcelable;
import android.preference.DialogPreference;
import android.util.AttributeSet;
import android.view.View;
import android.widget.NumberPicker;

import com.apps.kruszyn.lightorganapp.R;

/**
 * Created by nazyw on 4/23/2016.
 */
public class NumberPickerPreference extends DialogPreference {

    private static final int DEFAULT_VALUE = 0;

    private NumberPicker mNumberPicker;
    private Integer mCurrentValue;


    public NumberPickerPreference(Context context, AttributeSet attrs) {
        super(context, attrs);

        setDialogLayoutResource(R.layout.numberpicker_dialog);
        setPositiveButtonText(android.R.string.ok);
        setNegativeButtonText(android.R.string.cancel);

        setDialogIcon(null);
    }

    @Override
    protected void onBindDialogView(View view) {
        super.onBindDialogView(view);

        mNumberPicker = (NumberPicker)view.findViewById(R.id.numberPicker);
        mNumberPicker.setMinValue(0);
        mNumberPicker.setMaxValue(65535);

        if (mCurrentValue != null)
            mNumberPicker.setValue(mCurrentValue);
    }

    @Override
    protected void onDialogClosed(boolean positiveResult) {
        if (positiveResult) {
            mCurrentValue = mNumberPicker.getValue();
            persistInt(mCurrentValue);
        }
    }

    @Override
    protected void onSetInitialValue(boolean restorePersistedValue, Object defaultValue) {
        if (restorePersistedValue) {
            mCurrentValue = this.getPersistedInt(DEFAULT_VALUE);
        } else {
            mCurrentValue = (Integer) defaultValue;
            persistInt(mCurrentValue);
        }
    }

    @Override
    protected Object onGetDefaultValue(TypedArray a, int index) {
        return a.getInteger(index, DEFAULT_VALUE);
    }


    @Override
    protected Parcelable onSaveInstanceState() {
        final Parcelable superState = super.onSaveInstanceState();

        if (isPersistent()) {
            return superState;
        }

        final SavedState myState = new SavedState(superState);
        myState.value = mNumberPicker.getValue();
        return myState;
    }

    @Override
    protected void onRestoreInstanceState(Parcelable state) {

        if (state == null || !state.getClass().equals(SavedState.class)) {
            super.onRestoreInstanceState(state);
            return;
        }

        SavedState myState = (SavedState) state;
        super.onRestoreInstanceState(myState.getSuperState());

        mNumberPicker.setValue(myState.value);
    }


    private static class SavedState extends BaseSavedState {

        int value;

        public SavedState(Parcelable superState) {
            super(superState);
        }

        public SavedState(Parcel source) {
            super(source);
            value = source.readInt();
        }

        @Override
        public void writeToParcel(Parcel dest, int flags) {
            super.writeToParcel(dest, flags);
            dest.writeInt(value);
        }

        public static final Parcelable.Creator<SavedState> CREATOR =
                new Parcelable.Creator<SavedState>() {

                    public SavedState createFromParcel(Parcel in) {
                        return new SavedState(in);
                    }

                    public SavedState[] newArray(int size) {
                        return new SavedState[size];
                    }
                };
    }
}

File: Xamarin/LightOrganApp.Droid/UI/NumberPickerPreference.cs
using Android.Content;
using Android.OS;
using Android.Views;
using Android.Widget;
using Android.Preferences;
using Android.Util;
using Android.Content.Res;
using Java.Interop;
using Android.Runtime;

namespace LightOrganApp.Droid.UI
{
    [Register("com.apps.kruszyn.lightorganapp.droid.NumberPickerPreference")]
    public class NumberPickerPreference: DialogPreference
    {
        private const int DefaultValue = 0;

        private NumberPicker numberPicker;
        private int? currentValue;

        public NumberPickerPreference(Context context, IAttributeSet attrs): base(context, attrs)
        {
            DialogLayoutResource = Resource.Layout.numberpicker_dialog;
            SetPositiveButtonText(Android.Resource.String.Ok);
            SetNegativeButtonText(Android.Resource.String.Cancel);

            DialogIcon = null;
        }

        protected override void OnBindDialogView(View view)
        {
            base.OnBindDialogView(view);

            numberPicker = view.FindViewById<NumberPicker>(Resource.Id.numberPicker);
            numberPicker.MinValue = 0;
            numberPicker.MaxValue = 65535;
            
            if (currentValue != null)
                numberPicker.Value = currentValue.Value;
        }

        protected override void OnDialogClosed(bool positiveResult)
        {
            if (positiveResult)
            {
                currentValue = numberPicker.Value;
                PersistInt(currentValue.Value);
            }
        }
        
        protected override void OnSetInitialValue(bool restorePersistedValue, Java.Lang.Object defaultValue)
        {
            if (restorePersistedValue)
            {
                currentValue = GetPersistedInt(DefaultValue);
            }
            else
            {
                currentValue = (int)defaultValue;
                PersistInt(currentValue.Value);
            }
        }

        protected override Java.Lang.Object OnGetDefaultValue(TypedArray a, int index)
        {
            return a.GetInteger(index, DefaultValue);
        }

        protected override IParcelable OnSaveInstanceState()
        {
            var superState = base.OnSaveInstanceState();

            if (Persistent)
            {
                return superState;
            }

            var myState = new SavedState(superState);
            myState.Value = numberPicker.Value;
            return myState;
        }

        protected override void OnRestoreInstanceState(IParcelable state)
        {
            if (state == null || state.GetType() != typeof(SavedState)) {
                base.OnRestoreInstanceState(state);
                return;
            }

            var myState = (SavedState)state;
            base.OnRestoreInstanceState(myState.SuperState);

            numberPicker.Value = myState.Value;
        }

        public class SavedState: BaseSavedState
        {
            public int Value { get; set; }

            public SavedState(IParcelable superState): base(superState)
            {                
            }

            public SavedState(Parcel source): base(source)
            {                
                Value = source.ReadInt();
            }

            public override void WriteToParcel(Parcel dest, ParcelableWriteFlags flags)
            {
                base.WriteToParcel(dest, flags);
                dest.WriteInt(Value);
            }

            [ExportField("CREATOR")]
            static SavedStateCreator InitializeCreator()
            {
                return new SavedStateCreator();
            }

            class SavedStateCreator : Java.Lang.Object, IParcelableCreator
            {
                public Java.Lang.Object CreateFromParcel(Parcel source)
                {
                    return new SavedState(source);
                }

                public Java.Lang.Object[] NewArray(int size)
                {
                    return new SavedState[size];
                }
            }
        }          
    }
}